blocks/components/editable/format-toolbar.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * External dependencies
+ */
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+ // TODO: We mustn't import by relative path traversing from blocks to editor
+ // as we're doing here; instead, we should consider a common components path.
+import IconButton from '../../../editor/components/icon-button';
+
+const FORMATTING_CONTROLS = [
+	{
+		icon: 'editor-bold',
+		title: wp.i18n.__( 'Bold' ),
+		format: 'bold'
+	},
+	{
+		icon: 'editor-italic',
+		title: wp.i18n.__( 'Italic' ),
+		format: 'italic'
+	},
+	{
+		icon: 'editor-strikethrough',
+		title: wp.i18n.__( 'Strikethrough' ),
+		format: 'strikethrough'
+	}
+];
+
+class FormatToolbar extends wp.element.Component {
+	constructor( props ) {
+		super( ...arguments );
+		this.state = {
+			linkValue: props.formats.link ? props.formats.link.value : '',
+			isEditingLink: false
+		};
+		this.addLink = this.addLink.bind( this );
+		this.editLink = this.editLink.bind( this );
+		this.dropLink = this.dropLink.bind( this );
+		this.submitLink = this.submitLink.bind( this );
+		this.updateLinkValue = this.updateLinkValue.bind( this );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		const newState = {
+			linkValue: nextProps.formats.link ? nextProps.formats.link.value : ''
+		};
+		if (
+			! this.props.formats.link ||
+			! nextProps.formats.link ||
+			this.props.formats.link.node !== nextProps.formats.link.node
+		) {
+			newState.isEditingLink = false;
+		}
+		this.setState( newState );
+	}
+
+	toggleFormat( format ) {
+		return ( event ) => {
+			event.stopPropagation();
+			this.props.onChange( {
+				[ format ]: ! this.props.formats[ format ]
+			} );
+		};
+	}
+
+	addLink() {
+		if ( ! this.props.formats.link ) {
+			this.props.onChange( { link: { value: '' } } );
+
+			// Debounce the call to avoid the reset in willReceiveProps
+			setTimeout( () => this.setState( { isEditingLink: true } ) );
```

Comment:
Any time we use `setTimeout` or `debounce`, we can't be guaranteed that the component instance is still rendered when the callback is called. Therefore we must assign the timeout reference (`setTimeout` return value) into the instance and call `clearTimeout` in `componentWillUnmount`. Otherwise you'll encounter errors if the component is unmounted.

As a general note, this is one of the reasons I try to discourage `setTimeout` in a component (or any asynchronous action for that matter); because it's very difficult to track what's occurred in the time between when the timeout is declared and the callback is invoked.